src/aspire/storage/starfile.py

```diff
@@ -1,173 +1,134 @@
-                    if in_loop:
-                        field_names.append(line.split()[0])
+                    # assign key-value pair
+                    if gemmi_item.pair[0] not in pairs:
+                        pairs[gemmi_item.pair[0]] = gemmi_item.pair[1]
                     else:
-                        k, v = line.split()[:2]
-                        properties[k] = v
-
-                else:
-                    # we're looking at a data row
-                    tokens = line.split()
-                    if len(tokens) < len(field_names):
-                        logger.warning(
-                            f"Line {i} - Expected {len(field_names)} values, got {len(tokens)}."
+                        raise StarFileError(
+                            f"Duplicate key in pair: {gemmi_item.pair[0]}"
                         )
-                        tokens.extend([""] * (len(field_names) - len(tokens)))
-                    else:
-                        tokens = tokens[: len(field_names)]  # ignore any extra tokens
-
-                    rows.append(tokens)
-
-            # Any pending rows to be added?
-            if rows:
-                loops.append(pd.DataFrame(rows, columns=field_names, dtype=str))
-
-            # Any pending loops/properties to be added?
-            if loops or properties:
-                blocks.append(
-                    StarFileBlock(loops, name=block_name, properties=properties)
-                )
-
-            logger.info("StarFile parse complete")
-
-        logger.info("Initializing StarFile object from data")
-        self.init_from_blocks(blocks)
-        logger.info(f"Created <{self}>")
-
-    def init_from_blocks(self, blocks):
-        """
-        Initialize a StarFile from a list of blocks
-        :param blocks: An iterable of StarFileBlock objects
-        :return: An initialized StarFile object
-        """
-        for block in blocks:
-            self.blocks[block.name] = block
-
-    def __repr__(self):
-        return f"StarFile with {len(self.blocks)} blocks"
-
-    def __getitem__(self, item):
-        if isinstance(item, str):
-            return self.blocks[item]
-        else:
-            return self.blocks[list(self.blocks.keys())[item]]
+                if gemmi_item.loop is not None:
+                    block_has_loop = True
+                    if block_has_pair:
+                        raise StarFileError(
+                            "Blocks with multiple loops and/or pairs are not supported"
+                        )
+                    loop_tags = gemmi_item.loop.tags
+                    loop_data = [0 for x in range(gemmi_item.loop.length())]
+                    for row in range(gemmi_item.loop.length()):
+                        loop_data[row] = [
+                            gemmi_item.loop.val(row, col)
+                            for col in range(gemmi_item.loop.width())
+                        ]
+            if block_has_pair:
```

Comment:
No, this comes directly after iterating over the block. So firstly the Pair/Loop status of the block is identified (`block_has_pair` or `block_has_loop` and an error has been raised if they are both true), and in the same iteration the content has been collected (a dictionary of `Pair`s or a list of list representing the `loop`). So depending on whether it's a Pair or a Loop block, the DataFrame is constructed differently.

Reply:
Okay, probably you were going to add similar comments before I suggested we hold off on the comment until after first look.  This would def be some stuff we'd want to write down.